internal/api/v1/upload.go

```diff
@@ -59,26 +54,38 @@ func (hc ApplicationsController) Upload(w http.ResponseWriter, r *http.Request)
 		return InternalError(err, "failed to copy app sources to temp location")
 	}
 
-	log.V(2).Info("unpacking temp dir")
-	appDir := path.Join(tmpDir, "app")
-	err = archiver.Unarchive(blob, appDir)
+	cluster, err := kubernetes.GetCluster(ctx)
 	if err != nil {
-		return InternalError(err, "failed to unpack app sources to temp location")
+		return InternalError(err, "failed to get access to a kube client")
 	}
 
-	log.V(2).Info("create gitea app repo")
-	app := models.NewAppRef(name, org)
-	g, err := client.Upload(app, appDir)
+	connectionDetails, err := s3manager.GetConnectionDetails(ctx, cluster, deployments.TektonStagingNamespace, deployments.S3ConnectionDetailsSecret)
 	if err != nil {
-		return InternalError(err)
+		return InternalError(err, "fetching the S3 connection details from the Kubernetes secret")
+	}
+	manager, err := s3manager.New(connectionDetails)
+	if err != nil {
+		return InternalError(err, "creating an S3 manager")
 	}
 
-	log.Info("uploaded app", "org", org, "app", name)
+	username, err := GetUsername(r)
+	if err != nil {
+		return UserNotFound()
+	}
+	// TODO: When staging, check the metadata to ensure the blob that is requested
+	// actually belongs to the app
+	// TODO: We need the metadata to be able to tell if the blob belongs to an existing
+	// app or not (leftovers?).
```

Comment:
Let's also check if these metadata are really part of the S3 protocol or just a minio thing. In other words, can we be sure those data will always be there, regardless of the chose S3 solution?

Reply:
Moved this to a different issue. Needs discussion. 

https://github.com/epinio/epinio/issues/801